opal/mca/btl/tcp/btl_tcp_component.c

```diff
@@ -1121,97 +1117,64 @@ static int mca_btl_tcp_component_create_listen(uint16_t af_family)
-                     index = opal_ifnext(index)) {
-                 struct sockaddr_storage my_ss;
-
-                 /* Look if the module's address belongs to this
-                  * kernel IP interface.  If not, go to next address.
-                  */
-                 if (opal_ifindextokindex (index) !=
-                     mca_btl_tcp_component.tcp_btls[i]->tcp_ifkindex) {
-                     continue;
-                 }
-
-                 opal_ifindextoname(index, ifn, sizeof(ifn));
-                 opal_output_verbose(30, opal_btl_base_framework.framework_output,
-                                     "btl: tcp: component_exchange: examining interface %s",
-                                     ifn);
-                 if (OPAL_SUCCESS !=
-                     opal_ifindextoaddr(index, (struct sockaddr*) &my_ss,
-                                        sizeof (my_ss))) {
-                     opal_output (0,
-                             "btl_tcp_component: problems getting address for index %i (kernel index %i)\n",
-                             index, opal_ifindextokindex (index));
-                     continue;
-                 }
+     int rc;
+     size_t i;
+     size_t num_btls = mca_btl_tcp_component.tcp_num_btls;
+     size_t size = num_btls * sizeof(mca_btl_tcp_modex_addr_t);
+     mca_btl_tcp_modex_addr_t *addrs;
+
+     if (num_btls <= 0) {
+         return 0;
+     }
+
+     addrs = (mca_btl_tcp_modex_addr_t*)malloc(size);
+     if (NULL == addrs) {
+         return OPAL_ERR_OUT_OF_RESOURCE;
+     }
+     memset(addrs, 0, size);
+
+     for (i = 0 ; i < num_btls ; i++) {
+         struct mca_btl_tcp_module_t *btl = mca_btl_tcp_component.tcp_btls[i];
+         struct sockaddr *addr = (struct sockaddr*)&(btl->tcp_ifaddr);
 
 #if OPAL_ENABLE_IPV6
-                 if ((AF_INET6 == my_ss.ss_family) &&
-                     (6 != mca_btl_tcp_component.tcp_disable_family)) {
-                     memcpy(&addrs[current_addr].addr_inet,
-                             &((struct sockaddr_in6*)&my_ss)->sin6_addr,
-                             sizeof(addrs[0].addr_inet));
-                     addrs[current_addr].addr_port =
-                         mca_btl_tcp_component.tcp6_listen_port;
-                     addrs[current_addr].addr_family = MCA_BTL_TCP_AF_INET6;
-                     xfer_size += sizeof (mca_btl_tcp_addr_t);
-                     addrs[current_addr].addr_inuse   = 0;
-                     addrs[current_addr].addr_ifkindex =
-                         opal_ifindextokindex (index);
-                     current_addr++;
-                     opal_output_verbose(30, opal_btl_base_framework.framework_output,
-                                         "btl: tcp: component_exchange: "
-                                         "%s IPv6 %s", ifn,
-                                         opal_net_get_hostname((struct sockaddr*)&my_ss));
-                 } else
+         if (AF_INET6 == addr->sa_family) {
```

Comment:
We lost the check for mca_btl_tcp_component.tcp_disable_family (there is no need to pack an address that will never be used).

Reply:
Gah, you're right.  Will fix.